src/silly_sprites/opengl/shader_program.cpp

```diff
@@ -0,0 +1,179 @@
+                        auto message2 = error_message(id);
+                        spdlog::critical(
+                                "ERROR::SHADER::{}::COMPILATION_FAILED\nerror: {}",
+                                get_name_from_type(type),
+                                message2
+                        );
+                        glDeleteShader(id);
+                        throw GLError(
+                                GLErrorType::FailedToCompileVertexShader,
+                                fmt::format("first error: {}\nsecond error: {}", message, message2)
+                        );
+                    }
+                    break;
+
+                case Type::Geometry:
+                    spdlog::critical("ERROR::SHADER::{}::NO_FALLBACK", get_name_from_type(type));
+                    throw GLError(GLErrorType::FailedToCompileGeometryShader, message);
+                    break;
+
+                case Type::Fragment:
+                    success = compile_single(id, fallback_fragment_source);
+                    if (not success) {
+                        auto message2 = error_message(id);
+                        spdlog::critical(
+                                "ERROR::SHADER::{}::COMPILATION_FAILED\nerror: {}",
+                                get_name_from_type(type),
+                                message2
+                        );
+                        glDeleteShader(id);
+                        throw GLError(
+                                GLErrorType::FailedToCompileVertexShader,
+                                fmt::format("first error: {}\nsecond error: {}", message, message2)
+                        );
+                    }
+                    break;
+            }
+        }
+
+        spdlog::info("SUCCESS::SHADER::{}::COMPILATION", get_name_from_type(type));
+        return id;
+    }
+
+    void ShaderProgram::attach_shader(GLuint const shader) const {
+        glAttachShader(m_program_name, shader);
+        glDeleteShader(shader);
+    }
+
+    void ShaderProgram::link_program() const {
+        glLinkProgram(m_program_name);
+
+        GLint success;
+        glGetProgramiv(m_program_name, GL_LINK_STATUS, &success);
+        if (not success) {
+            auto message = std::string{};
+            GLint len;
+            glGetProgramiv(m_program_name, GL_INFO_LOG_LENGTH, &len);
+            message.resize(gsl::narrow<usize>(len - 2));
+            glGetProgramInfoLog(m_program_name, len - 1, nullptr, message.data());
+            spdlog::critical("ERROR::PROGRAMM::LINK_FAILED -> {}\n", message.data());
+            glDeleteProgram(m_program_name);
+            throw GLError(GLErrorType::FailedToLinkShaderProgram, message);
+        }
+    }
+
+    ShaderProgram::ShaderProgram(
+            std::string_view const vertex_source,
+            std::string_view const geometry_source,
+            std::string_view const fragment_source
+    )
+        : m_program_name{ 0 } {
```

Comment:
Was this done by clang format? It looks weird. But maybe just don't use the initializer list to initialize `m_program_name` and set a default value in the class itself.
```suggestion
    ) : m_program_name{ 0 } {
```